freedom-zephyrdtsfixup-generator.c++

```diff
@@ -0,0 +1,145 @@
+/* Copyright 2018 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "fdt.h++"
+#include "libfdt.h++"
+#include <fstream>
+#include <iostream>
+#include <string>
+
+using std::cerr;
+using std::endl;
+using std::fstream;
+using std::string;
+
+static void show_usage(string name)
+{
+  std::cerr << "Usage: " << name << " <option(s)>\n"
+	    << "Options:\n"
+	    << "\t-h,--help\t\t\tShow this help message\n"
+	    << "\t-d,--dtb <eg. xxx.dtb>\t\tSpecify fullpath to the DTB file\n"
+	    << "\t-o,--output <eg. dts.fixup>\t\tGenerate Zephyr RTOS DTS fixup file\n"
+	    << endl;
+}
+
+static void write_config_file(const fdt &dtb, fstream &os)
+{
+  os << "/* Copyright (c) 2018 SiFive Inc." << endl;
+  os << " * SPDX-License-Identifier: Apache-2.0 */" << endl;
+
+  auto emit_comment = [&](const node &n, string dev_name, int dev_num) {
+    os << endl << "/* " << dev_name << "_" << dev_num << " -> " << n.name() << " */\n";
+  };
+
+  auto emit_fixup = [&](string out_prefix, string in_prefix, int dev_num, target_addr base, string suffix) {
+    os << "#define ";
+    os << out_prefix << "_" << dev_num << "_" << suffix << "\t";
+    os << in_prefix << "_" << std::hex << base << std::dec << "_" << suffix << endl;
```

Comment:
This is an aside, but is there any way to emit an integer as hex to a stream and then reset it to it's original state?  I always use this idiom, but it assumes the stream was in std::dec mode.  I feel like @brucehoult might know...

Reply:
No standard way using manipulators inline in a series of << no.

There are a bunch of suggestions here, but still none that are very pretty: https://stackoverflow.com/questions/2273330/restore-the-state-of-stdcout-after-manipulating-it